File: Cores/JVM/src/main/scala/OSOL/Extremum/Cores/JVM/Optimization/RemoteFunctions/RemoteFunction.scala
package OSOL.Extremum.Cores.JVM.Optimization.RemoteFunctions

import scala.sys.process.Process

abstract class RemoteFunction[FuncType](val json: String, val port: Int, val field: String) {

  val process_specs = Process("run_core", Seq("--core", json, "--port", port.toString))
  var server_process: Process = null

  def initialize(): Unit = {
    server_process = process_specs.run()
    Thread.sleep(5000)
  }

  def apply(values: Map[String, FuncType]): FuncType

  def terminate(): Unit = {
    server_process.destroy()
  }

}

File: Cores/DotNet/Optimization/RemoteFunctions/RemoteFunction.cs
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.Net.Cache;
using System.Net.Http;

namespace OSOL.Extremum.Cores.DotNet.Optimization.RemoteFunctions
{
    public abstract class RemoteFunction<TFuncType>
    {
        public string Json;
        public int Port;
        public string Field;
        private ProcessStartInfo ScriptConfig;
        public Process ServerProcess;
        public WebClient Client = new WebClient();

        protected RemoteFunction(string json, int port, string field)
        {                
            this.Json = json;
            this.Port = port;
            this.Field = field;
            
            this.ScriptConfig = new ProcessStartInfo();
            this.ScriptConfig.FileName = "run_core";
            this.ScriptConfig.Arguments = $"--core {this.Json} --port {this.Port}";
            this.ScriptConfig.WindowStyle = ProcessWindowStyle.Hidden;
            
            this.Client.Proxy = GlobalProxySelection.GetEmptyWebProxy();
            this.Client.BaseAddress = $"http://localhost:{this.Port}";
            this.Client.Headers = null;
        }

        public void Initialize()
        {
            ServerProcess = Process.Start(ScriptConfig);
            System.Threading.Thread.Sleep(5000);
        }

        public abstract TFuncType Calculate(Dictionary<string, TFuncType> values);

        public void Terminate()
        {
            ServerProcess.Kill();
        }
    }
}